### PR TITLE
📖 document exposed metrics

### LINKED ISSUE #262: Document what metrics are registered in godoc
From https://github.com/kubernetes-sigs/controller-runtime/pull/234#issuecomment-448248502

### PR DESCRIPTION
**Pull request by djzager**
Update documentation for the metrics package to enumerate all exposed
metrics with an explanation of what they capture.

Fixes #262

### CODE REVIEW

**Comment by djzager** [comment]
I see that the [original comment](https://github.com/kubernetes-sigs/controller-runtime/pull/234#issuecomment-448248502) was referring to the `pkg/webhook/internal/metrics` package but it already documents the exposed metrics.

 I could follow the same pattern as [`pkg/webhook/internal/metrics`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/internal/metrics/metrics.go#L26-L28) but that would require exporting a bunch of stuff that is currently not exported.

**Comment by djzager** [comment]
Looking around the project at the other metrics packages (ie. [controller metrics](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/internal/controller/metrics) and [webhook metrics](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/webhook/internal/metrics)) I feel that I took the wrong approach in this PR.

What _I would like to do_ would be to export the variables in this package where metrics are defined (making them `var` where they are now hidden inside a function) and add the appropriate comments.

@DirectXMan12 @vincepri WDYT?

**Comment by vincepri** [comment]
That sounds good to me as well

File: pkg/metrics/workqueue.go [hunk 1]
```diff
@@ -19,106 +19,118 @@ package metrics
...
 
-var log = logf.RuntimeLog.WithName("metrics")
-
 // This file is copied and adapted from k8s.io/kubernetes/pkg/util/workqueue/prometheus
 // which registers metrics to the default prometheus Registry. We require very
 // similar functionality, but must register metrics to a different Registry.
 
+var (
+	// WorkQueueSubsystem defines the subsystem for the workqueue metrics.
+	WorkQueueSubsystem = "workqueue"
```
Comment (djzager) [inline pkg/metrics/workqueue.go]: went ahead and followed what was being done in [kubernetes metrics](https://github.com/kubernetes/kubernetes/blob/c7d84ac77e3064c80c7f9c9e11a017b0bf78acd9/staging/src/k8s.io/component-base/metrics/prometheus/workqueue/metrics.go)

**Comment by djzager** [comment]
Went ahead and left it as 2 commits to make it easier to go back to the original if necessary. If this works I'll rebase and squash.

**Comment by DirectXMan12** [comment]
Hey, so I'm sorry to jump in a bit later here, but: 

> What I would like to do would be to export the variables in this package where metrics are defined (making them var where they are now hidden inside a function) and add the appropriate comments.

I'm not a huge fan of this approach.  Namely, it means we have out metrics compatibility boundary not at the metrics we expose, but via the *exact* method we expose them by.  For instance, under this model, switching to OpenTelemetry is *necessarily* a breaking change, even if we expose the same metrics in the same format with the same names, since the Go types change.  Considering that the Go types are mostly an implementation detail, that seems suboptimal.

As I see it, there are two ways around this: move the actual variables to an internal package so that users can't depend on them, or make the variables private and document as a block.

Thoughts?

/hold

(while we're discussing)

**Comment by djzager** [comment]
> Due to the way this is currently structured, it means we have out metrics compatibility boundary not at the metrics we expose, but via the _exact_ method we expose them by. For instance, under this model, switching to OpenTelemetry is _necessarily_ a breaking change, even if we expose the same metrics in the same format with the same names, since the Go types change. Considering that the Go types are mostly an implementation detail, that seems suboptimal.

My idea of making unexported `var`s exported made me uneasy, but I didn't really understand why. Thank you for the explanation.

I have updated the metrics package to expose the subsystems and specific keys used without exporting the metrics themselves.

I would be tempted to make similar changes in https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/internal/metrics/metrics.go and https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/metrics/metrics.go but I suspect making exported things unexported would also be a breaking change. It's also possible that `pkg/internal/controller/metrics` and `pkg/webhook/internal/metrics` are fundamentally different than `pkg/metrics` but I'm not certain. Hoping to gain better understanding of what changes make sense and when.

**Comment by DirectXMan12** [comment]
> It's also possible that pkg/internal/controller/metrics and pkg/webhook/internal/metrics are fundamentally different than pkg/metrics but I'm not certain

Ah, yes, your intuition is correct here :-).  In Go, and package with `/internal/` in the path as visibility limited to the parent package of `/internal` (as well as any subpackages of the parent package).  So anything in `/pkg/webhook/internal/...` can only be accessed by things in `/pkg/webhook/...`.  This is enforced by the compiler's import logic: https://golang.org/doc/go1.4#internalpackages

**Comment by shawn-hurley** [comment]
Should we cancel the hold on this?

**Pull request merged**